### DISCUSSION TITLE
Jenkins plugins datasource versioning is broken

### DISCUSSION THREAD

**Discussion by aabouzaid** [renovatebot/renovate#13793] [Suggest an Idea]
### How are you running Renovate?

WhiteSource Renovate hosted app on github.com

### Describe the bug

Jenkins rolled out a new CI flow to speed up the Jenkins plugin release. (more details in: [JEP-305](https://github.com/jenkinsci/jep/blob/master/jep/305/README.adoc), and [JEP-229](https://github.com/jenkinsci/jep/blob/master/jep/229/README.adoc)) So a new versioning format is used instead of SemVer

However, the format is not unified anymore and each plugin is using a bit different format! Till now I detected 4 formats:
- 82.4facab766924
- 336.v182c0fbaaeb7 (e.g. [hashicorp-vault-plugin](https://plugins.jenkins.io/hashicorp-vault-plugin/))
- 2.13.0-230.v59243c64b0a5 (e.g. [jackson2-api](https://plugins.jenkins.io/jackson2-api/))
- 2.8-138.d766e30bb08b (e.g. [github-api](https://plugins.jenkins.io/github-api/))

That's in addition to the plugins that use the normal SemVer.

And since [Jenkins plugins datasource](https://github.com/renovatebot/renovate/blob/cc76e41/lib/datasource/jenkins-plugins/index.ts) in Renovate doesn't have its own method to handle the versions, thus it uses the default (which is 'SemVer') and it not working anymore. The Jenkins plugins upgrade PR is broken because some plugins have been upgraded and some have not.

So no now Jenkins plugins datasource should get its own versioning method [like many other datasources](https://github.com/renovatebot/renovate/tree/cc76e41/lib/versioning).

### Have you created a minimal reproduction repository?

Yes
https://github.com/aabouzaid/poc-renovatebot-jenkins-plugins-bug-13373/pull/2

In PR there are 2 Jenkins plugins, one that still uses SemVer (htmlpublisher) and another uses the unsupported versioning format (jackson2-api).

The plugin `htmlpublisher` will be updated from `1.25` to `1.28` which is the latest release according to the plugin page:
https://plugins.jenkins.io/htmlpublisher/#releases

However, The plugin `jackson2-api` will NOT be updated to the latest release it will be only updated from `2.12.3` to `2.12.4` which is the latest release that uses SemVer, but the actual latest release is `2.13.1-244.v773c36c5b330`:
https://plugins.jenkins.io/jackson2-api/#releases

**Comment by aabouzaid** [comment]
@rarkins I've created a reproduction repository and updated the issue with the details.

**Comment by viceice** [comment]
I think you can workaround via assign different versioning with package rules.

**Comment by aabouzaid** [comment]
@viceice is it possible with all those version variates?

**Comment by rarkins** [comment]
If it's not possible to assign the right versioning via packageRules then there will be nothing which the Jenkins manager can do about it either. 

I suggest that in the Jenkins manager we can test if the current value is valid semver and if not then fall back to loose versioning. If any other versioning is needed then the user can apply it via packageRules, assuming we can't easily auto detect it.

**Comment by aabouzaid** [comment]
@rarkins as I understand (and please correct me if I'm wrong), each datasource could set its own way to handle the versioning as shown in the [versioning](https://github.com/renovatebot/renovate/tree/cc76e41/lib/versioning) folder and if that is not set [it uses the default value](https://github.com/renovatebot/renovate/blob/ea28c6a/lib/datasource/index.ts#L213) "SemVer" by default.

So all that's needed is to create the version handler for Jenkins plugins which simply handles all combinations of those version formats and use it by default.

**Comment by rarkins** [comment]
> each datasource could set its own way to handle the versioning as shown in the [versioning](https://github.com/renovatebot/renovate/tree/cc76e41/lib/versioning?rgh-link-date=2022-01-08T05%3A19%3A26Z) folder and if that is not set [it uses the default value](https://github.com/renovatebot/renovate/blob/ea28c6a/lib/datasource/index.ts?rgh-link-date=2022-01-08T05%3A19%3A26Z#L213) "SemVer" by default.

Almost. Each datasources has a *default* versioning if none is specified. But if the manager specifies both a datasource and a versioning value then it will be used and not the datasource's default.

> So all that's needed is to create the version handler for Jenkins plugins which simply handles all combinations of those version formats and use it by default.

Sorry, but I don't understand what you mean by this.

As a workaround, you can use `packageRules` to override the `versioning` for any extracted dependency (e.g. match using `matchPackageNames`).

As a full solution which doesn't require manual configuration, the Jenkins manager would need to "guess" which versioning is in use at time of extraction and then assign that as part of its extraction result. e.g. you can see an example here where the npm manager specifies `node` versioning for this specific case: https://github.com/renovatebot/renovate/blob/013189bf1840762b4754a21dab76dc18d7709890/lib/manager/npm/extract/index.ts#L191-L195

**Comment by timja** [comment]
@rarkins there is a reproduction

**Comment by viceice** [comment]
Those versions seem to be compatible with our maven versioning.

As those plugins are build by maven.

So i think we can simply assign maven as default versioning to Jenkins plugins.

**Reply by timja** [reply]
Yes it is compatible with Maven

**Comment by rarkins** [comment]
Converted this to a Discussion, as it is misfiled as a bug and can be solved with config too.

**Comment by maxdanilov** [comment]
The default versioning scheme for Jenkins manager changed to `maven` here: ~https://github.com/renovatebot/renovate/discussions/13793~ https://github.com/renovatebot/renovate/pull/14242

**Reply by timja** [reply]
Thanks! looks to be https://github.com/renovatebot/renovate/pull/14242 actually ❤️ 

Will keep an eye on it

**Reply by timja** [reply]
Working, thanks! 
https://github.com/hmcts/cnp-jenkins-docker/pull/572
